tower-http/src/compression/body.rs

```diff
@@ -1,320 +1,233 @@
-            };
-
-        match bytes {
-            Some(Ok(data)) => Poll::Ready(Some(Ok(BytesMut::freeze(data)))),
-            Some(Err(err)) => {
-                let err = inner
-                    .project()
-                    .error
-                    .take()
-                    .map(Error::Body)
-                    .unwrap_or(Error::Compress(err));
-                Poll::Ready(Some(Err(err)))
-            }
-            None => Poll::Ready(None),
+                Some(Err(err)) => Poll::Ready(Some(Err(err))),
+                None => Poll::Ready(None),
+            },
         }
     }
 
     fn poll_trailers(
         self: Pin<&mut Self>,
         cx: &mut Context<'_>,
-    ) -> Poll<Result<Option<http::HeaderMap>, Self::Error>> {
-        match self.project().inner.project() {
+    ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
+        match self.project().0.project() {
             #[cfg(feature = "compression-gzip")]
-            BodyInnerProj::Gzip(framed_read) => framed_read
-                .get_pin_mut()
-                .get_pin_mut()
-                .get_pin_mut()
-                .project()
-                .body
-                .poll_trailers(cx),
+            BodyInnerProj::Gzip(inner) => inner.poll_trailers(cx),
             #[cfg(feature = "compression-deflate")]
-            BodyInnerProj::Deflate(framed_read) => framed_read
-                .get_pin_mut()
-                .get_pin_mut()
-                .get_pin_mut()
-                .project()
-                .body
-                .poll_trailers(cx),
+            BodyInnerProj::Deflate(inner) => inner.poll_trailers(cx),
             #[cfg(feature = "compression-br")]
-            BodyInnerProj::Brotli(framed_read) => framed_read
-                .get_pin_mut()
-                .get_pin_mut()
-                .get_pin_mut()
-                .project()
-                .body
-                .poll_trailers(cx),
-            BodyInnerProj::Identity(inner, _) => inner.poll_trailers(cx),
+            BodyInnerProj::Brotli(inner) => inner.poll_trailers(cx),
+            BodyInnerProj::Identity(body) => body.poll_trailers(cx),
         }
-        .map_err(Error::Body)
     }
 }
 
-#[pin_project(project = BodyInnerProj)]
-#[derive(Debug)]
-enum BodyInner<B, E> {
-    #[cfg(feature = "compression-gzip")]
-    Gzip(#[pin] FramedRead<GzipEncoder<StreamReader<Adapter<B, E>, Bytes>>, BytesCodec>),
-    #[cfg(feature = "compression-deflate")]
-    Deflate(#[pin] FramedRead<ZlibEncoder<StreamReader<Adapter<B, E>, Bytes>>, BytesCodec>),
-    #[cfg(feature = "compression-br")]
-    Brotli(#[pin] FramedRead<BrotliEncoder<StreamReader<Adapter<B, E>, Bytes>>, BytesCodec>),
-    Identity(#[pin] B, PhantomData<fn() -> E>),
-}
-
-#[pin_project]
-#[derive(Debug)]
-struct Adapter<B, E> {
-    #[pin]
-    body: B,
-    error: Option<E>,
-}
+#[cfg(feature = "compression-gzip")]
+impl<B> DecorateAsyncRead for GzipEncoder<B>
+where
+    B: Body,
+    B::Error: Into<io::Error>,
```

Comment:
This bound won't work for hyper? https://docs.rs/hyper/0.14.4/hyper/struct.Body.html#impl-Body

Reply:
Yes. The error is converted into an `io::Error` using `io::ErrorKind::Other` [here](https://github.com/tower-rs/tower-http/blob/dedup-compression-and-decompression/tower-http/src/compression_utils.rs#L301).

Just to be sure I added a test in 788390426b175.

Reply:
I am not sure I like that, why not create a specific error type so that a user can potentially downcast to that type instead of having to string match on an io error?

Reply:
Yeah I agree it’s not great. The fact that AsyncRead requires using io::Error complicates things a bit since Body can have any error type.

The original version of this did actually have a worker around where it would store an error of another type and keep a sentinel io::Error that would be converted later.

I’ll give it another look 👀

Reply:
I found a fix. 726144bae566d.

Added [`StreamErrorIntoIoError`](https://github.com/tower-rs/tower-http/blob/dedup-compression-and-decompression/tower-http/src/compression_utils.rs#L291) which converts errors from a stream into a sentinel `io::Error` and allowing you to get out the actual error later.

So `CompressionBody` and `DecompressionBody` now both use [`BodyOrIoError`](https://github.com/tower-rs/tower-http/blob/dedup-compression-and-decompression/tower-http/src/lib.rs#L208) as their error type and so retains the original errors from the `Body`.